ffbidx indexer

hcstadler opened the pull request:
Add the fast feedback indexer (https://github.com/paulscherrerinstitute/fast-feedback-indexer), an implementation of the Toro algorithm described at  https://journals.iucr.org/j/issues/2024/04/00/jo5098/index.html

Use --help-ffbidx to see options and --indexing=ffbidx to use it with indexamajig.

pmiddend commented:
Can we close #14 now?

fleon-psi commented:
> Can we close #14 now?

Yes :), let's continue with #15

hcstadler commented on libcrystfel/src/crystfel-mille.c:
This diff has nothing to do with ffbidx, as far as I can tell. This diff must therefore come from Tom's fork-exec branch, probably.

hcstadler commented on src/im-sandbox.c:
This diff is not directly related to ffbidx. I guess it comes from Tom's fork-exec branch that served as the basis for this pull request.

hcstadler commented:
I have dropped the changes unrelated to ffbidx.

pmiddend commented:
Thanks! I've tried compiling this, but I get:

```
../libcrystfel/src/indexers/ffbidx.c: In function ‘ffbidx_prepare’:
../libcrystfel/src/indexers/ffbidx.c:254:54: error: ‘MEMORY_PIN_STATIC’ undeclared (first use in this function)
  254 |     priv->handle = create_indexer(&priv->opts.cpers, MEMORY_PIN_STATIC, &priv->err_msg, NULL);
      |                                                      ^~~~~~~~~~~~~~~~~
```

I'm not sure where this identifier comes from. Did I do something wrong?

hcstadler commented:
pmiddend,

this constant was introduced recently, so most likely you will be ok if you do a git pull on the fast feedback indexer code, recompile and install.

This is related to memory pinning, static means to keep a permanent pinned buffer. This safes on memory allocations/deallocations and pinning operations. Pinning fixes the virtual to physical memory translation table for a piece of memory and speeds up Host<->GPU data transfer.

Let me know if you still have problems with the newest code.

pmiddend commented:
Thanks @hcstadler, I've been able to compile and run the branch on some test data.

Sorry in advance that this is premature, I haven't had time for more debugging, but maybe you can already give hints: I've tried this with a random `-j` parallelization value of 4 (on a V100 and A100 GPU). This resulted in a frame rate of <10fps, which is a bit slower than I expected. I tried to increase the parallelization, but ended up with `Error creating indexer` and CrystFEL bailing out early.  Am I doind something wrong? Is there a `-j` sweet spot, or can I get more information on the indexer creation error?

hcstadler commented:
@pmiddend: The most likely cause for such an error is missing space on the GPU, especially if you should have compiled the ffbidx library without giving meson the --buildtype=release argument. Contact me at hans-christian.stadler@psi-ch for further discussions on this, unless you see a connection to this pull request. The current limit is given by GPU memory and some parameters to ffbidx with an influence on the size of the indexer state on the GPU.  On my laptop I used -j 32 and the default ffbidx parameter settings, which resulted in a memory usage less than half of the 8GB available.
There is a sweet spot for ffbidx which is around 4 on my laptop, but the overall sweet spot for indexamajig is much higher, as ffbidx is not the bottleneck.

hcstadler commented:
I had to rebase everything on the master branch in order to prevent messing up the git history (my code was originally based on some version of Toms fork-exec branch).

taw10 commented:
Sorry for keeping you waiting.  I think it's pretty close to being merged, but here are two things I noticed:

First, please make sure it compiles without the ffbidx library.  For `pinkindexer`, we simply wrap the entire source file with `#ifdef HAVE_PINKINDEXER` and add dummy routines at the bottom.  You could also disable compilation of ffbidx.c in meson.build - see for example how `indexamajig` handles a missing `zmqdep` in the top-level meson.build.

Second, please make sure that the indentation style is consistent within already-existing files.  In C files, that's "tabs for indentation, spaces for alignment".  Meson files are spaces only.  You don't have to re-format `ffbidx.c` and `ffbxidx.h` however, since I'm thinking of changing the format project-wide in the medium term future.

I'll reply separately about the related things you asked by email!

hcstadler commented:
Thanks Tom for pointing out these deficiencies. Hope I now fixed the spacing everywhere and masked out code aking to the pinkIndexer (as Filip did originally) so that it now compiles without ffbidx. I also rebased on the main branch.

hcstadler commented:
Again I rebased on master and squashed all commits into one to cleanup git history (I have to stop using the update button for the fork in the github gui ;-))

hcstadler commented:
I looks like the plugin stopped functioning properly. I'll investigate...

hcstadler commented:
False alarm, stupid mistakes from my side.

hcstadler commented:
Squashed commits into one.

hcstadler commented:
Thanks to Toms testing revealing some deficiencies, I added an indexing precondition check on the number of minimal spots required, and a cell viability check. I hope this addresses the deficiencies found.

taw10 commented:
Merged.  Thanks for all your hard work!
